### PR TITLE
[Model] Add BGNN example

### PR DESCRIPTION
**Pull request by nd7141**
## Description
A proposal of a new GNN model coming from ICLR 2021: https://openreview.net/forum?id=ebS5NUfoMKL
The model is implemented with DGL: https://github.com/nd7141/bgnn

Reviewers: @jermainewang @BarclayII 

Discussion thread: https://github.com/dmlc/dgl/issues/2603

This is initial proposal of the model. There are two files: `BGNN.py` which implements BGNN model and `run.py` which shows how to use it. 

I also include datasets I used in the paper for the sake of reproducibility (although I think that datasets should be included in DGL as a separate request). 

Current results match those of the papers (although I have not done thorough hyperparameter tuning). 

Here are images showing performance of BGNN, that is similar to the paper (https://openreview.net/forum?id=ebS5NUfoMKL): 
image
image
image
image
image
image

I think `BGNN.py` should be part of the DGL, in the sense that an end user can just do `import dgl.BGNNPredictor`. On the other hand, `run.py` can be kept in the example. 

A few notes: 
* Currently `BGNN.py` includes as well plotting, which may not be needed for the final version.
* I evaluate for train, valid, and test masks during training, which can slow down training. These parts could be excluded, but should be discussed as it impacts early stopping. 
* BGNN relies on CatBoost package (https://catboost.ai/), so it should be installed if an end user uses BGNN. 
* There are some packages which could be installed as well but are not necessary (e.g. tqdm, sklearn). 

I am happy to answer your questions and discuss how to integrate BGNN in DGL in the optimal way. Please let me know your thoughts.

### CODE REVIEW

**Comment by BarclayII** [comment]
Thanks a lot!  It's a brand new model contribution so I put an item in the model index README.

Could you also add a README with instructions on how to download the dataset (once you have uploaded them)?

**Comment by nd7141** [comment]
@BarclayII @jermainewang I updated the code in the example and provided instructions how you can download datasets. Please let me know if something else is needed from me?

File: examples/pytorch/bgnn/BGNN.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+                 gnn_model,
+                 task = 'regression',
+                 loss_fn = None,
+                 trees_per_epoch = 10,
+                 backprop_per_epoch = 10,
+                 lr=0.01,
+                 append_gbdt_pred = True,
+                 train_input_features = False,
+                 gbdt_depth=6,
+                 gbdt_lr=0.1):
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:25]: Could you pls add docstring for these? We follow the [numpy doc style](https://numpydoc.readthedocs.io/en/latest/format.html). [Code example in DGL](https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/graphconv.py#L141). It's good to have some code examples for usage.

File: examples/pytorch/bgnn/BGNN.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+                catboost_loss_fn = 'MultiClass'
+            else: # we predict the gradients for each class at epochs > 0
+                catboost_model_obj = CatBoostRegressor
+                catboost_loss_fn = 'MultiRMSE'
+
+        return catboost_model_obj(iterations=num_epochs,
+                                  depth=self.gbdt_depth,
+                                  learning_rate=self.gbdt_lr,
+                                  loss_function=catboost_loss_fn,
+                                  random_seed=0,
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:57]: expose `random_seed` to the init function of `BGNNPredictor`?

File: examples/pytorch/bgnn/BGNN.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+        if self.task == 'regression':
+            self.out_dim = y.shape[1]
+        elif self.task == 'classification':
+            self.out_dim = len(set(y.iloc[test_mask, 0]))
+        self.in_dim = self.out_dim + X.shape[1] if self.append_gbdt_pred else self.out_dim
+
+        gbdt_X_train = X.iloc[train_mask]
+        gbdt_y_train = y.iloc[train_mask]
+        gbdt_alpha = 1
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:268]: Is this a tunable hyperparameter? Consider expose it to the `__init__` function with default value.

File: examples/pytorch/bgnn/BGNN.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            train_mask, val_mask, test_mask,
+            cat_features = None,
+            num_epochs=100,
+            patience=10,
+            logging_epochs=1,
+            metric_name='loss',
+            normalize_features=False,
+            replace_na=True,
+            ):
+        '''
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:230]: Same. Pls follow the numpy docstring style.  DGL also has some conventions for docing the argument types:
* Use `dgl.DGLGraph` for DGLGraph.
* Use `numpy.array` for numpy arrays and describe the shape in the description.
* Use `list[int]` for list of integers
* Use python type names for python objects (e.g., `int`, `float`, ...)

File: examples/pytorch/bgnn/BGNN.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+        :param graph: dgl graph
+        :param X: numpy matrix of input features. Can be composed of categorical features with missing values.
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:233]: describe how missing values are represented in the matrix

File: examples/pytorch/bgnn/BGNN.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        :param y: numpy matrix of target labels. Can be multidimensional targets.
+        :param train_mask: list of integers (rows in X that correspond to train set)
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:235]: describe the element of the mask array (e.g., 0-1 array?)

File: examples/pytorch/bgnn/BGNN.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        :param val_mask: list of integers (rows in X that correspond to valid set)
+        :param test_mask: list of integers (rows in X that correspond to test set)
+        :param cat_features: list of integers (columns in X that correspond to categorical features)
+        :param num_epochs: number of epochs to run
+        :param patience: number of epochs to wait until early stopping
+        :param logging_epochs: log every n epoch
+        :param metric_name: metric to use for early stopping
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:242]: I wonder how different early stopping metric performs in practice? Is it critical? Can we remove the option and use the default value `'loss'`?

File: examples/pytorch/bgnn/BGNN.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+        :param normalize_features: normalize original input features X (column wise)
+        :param replace_na: replace NA in X
```
Comment (jermainewang) [inline examples/pytorch/bgnn/BGNN.py:244]: I wonder if these options (`normalize_features` and `replace_na`) should be here. They look like data preprocessing steps that may go to the example code.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: examples/pytorch/bgnn/BGNN.py (hunk 1 above)
Comment (nd7141) [reply examples/pytorch/bgnn/BGNN.py]: Updated.

File: examples/pytorch/bgnn/BGNN.py (hunk 7 above)
Comment (nd7141) [reply examples/pytorch/bgnn/BGNN.py]: I think we better keep as it is, as we want to look at accuracy and not at the loss, if accuracy if our final goal.

File: examples/pytorch/bgnn/BGNN.py (hunk 8 above)
Comment (nd7141) [reply examples/pytorch/bgnn/BGNN.py]: Moved them to run.py for preprocessing.

**Comment by nd7141** [comment]
Hey @jermainewang 

I updated the code documentation and also moved the preprocessing to run.py (that requires adding one additional parameter for fit function, as GBDT still want to operate on non-preprocessed features). That worked on my side. I hope it works for you too.

**Review by jermainewang — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by jermainewang** [comment]
@BarclayII do you have other comments?

**Comment by BarclayII** [comment]
I'm good.  Thanks!
